### PR TITLE
[aot_compile] Add training and distributed DTensor documentation

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184230

Add a Training section to the aot_compile docs covering:

- Basic training with aot_compile (forward+backward captured automatically
  when inputs require gradients, save/load preserves autograd support)
- Distributed training with DTensor using compile_on_one_rank to produce
  rank-independent compiled graphs, with a complete torchrun example

All code examples verified on 8x H100 GPUs. Verification scripts and output:
https://gist.github.com/bobrenjc93/68ac4382a3bc119c394af4648a118741

Authored by Claude.

### CODE REVIEW

File: docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md [hunk 1]
```diff
@@ -217,6 +217,157 @@ with open("fn_with_model.pt", "rb") as f:
     )
 ```
 
+## Training
```
Comment (aorenste) [inline docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md:222]: This section's first example largely duplicates the 'Compiling a module' section above (lines 65-110): same `MyModel`, same `torch.compile(model, fullgraph=True).forward.aot_compile(((torch.randn(3, 4, requires_grad=True),), {}))`, same save/load pattern. The genuinely new content is the optimizer + training loop. Consider trimming this example down to just the optimizer-loop delta (with a reference back to 'Compiling a module'), or folding the training-loop content into that earlier section so the docs have one canonical place that covers train-loop usage.

File: docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md [hunk 2]
```diff
@@ -217,6 +217,157 @@ with open("fn_with_model.pt", "rb") as f:
...
+# AOT compile -- backward is captured automatically when inputs require grad.
+compiled_fn = torch.compile(
+    model, fullgraph=True
+).forward.aot_compile(((torch.randn(3, 4, requires_grad=True),), {}))
+
+# Training loop using the AOT-compiled function.
+optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
+for _ in range(3):
+    optimizer.zero_grad()
+    inputs = torch.randn(3, 4, requires_grad=True)
```
Comment (aorenste) [inline docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md]: Caveat: the pre-existing 'Compiling a module' example (lines 88, 91, 106) already uses `requires_grad=True` on raw `torch.randn` inputs, so this PR is propagating the pattern rather than introducing it. That said, since this PR is *adding* new examples, it's a natural opportunity to set a better one -- real training data never has `requires_grad=True`, and the prose at line 222 even says 'When any input **or parameter** requires gradients...', so `model.parameters()` alone suffice to capture backward.

Ideally: drop `requires_grad=True` from both the example input (line 245) and the training-loop inputs here (line 251), and update the existing 'Compiling a module' example to match. At minimum, drop it from the new content so readers see at least one correct pattern. As written, anyone copying this will conclude they need to mark dataloader tensors `requires_grad=True`.

File: docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md [hunk 3]
```diff
@@ -217,6 +217,157 @@ with open("fn_with_model.pt", "rb") as f:
...
+
+with open("train_model.pt", "rb") as f:
+    loaded_fn = torch.compiler.load_compiled_function(f)
+
+inputs = torch.randn(3, 4, requires_grad=True)
+output = loaded_fn(model, inputs)
+output.sum().backward()  # gradients flow correctly
+```
+
+(distributed-training-with-dtensor)=
```
Comment (aorenste) [inline docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md]: This anchor isn't referenced anywhere in `docs/`. Either add a `{ref}` to it from somewhere that should cross-link here, or remove the anchor.

File: docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md [hunk 4]
```diff
@@ -217,6 +217,157 @@ with open("fn_with_model.pt", "rb") as f:
...
+    # Training loop.
+    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
+    for step in range(5):
+        optimizer.zero_grad()
+        x = DTensor.from_local(
+            torch.randn(4, 64, device=f"cuda:{rank}"),
+            mesh, [Replicate()], run_check=False,
+        )
+        out = compiled_fn(model, x)
+        local_out = out.to_local() if isinstance(out, DTensor) else out
```
Comment (aorenste) [inline docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md]: This defensive runtime check is never motivated in the surrounding prose. The section above claims `compile_on_one_rank` produces 'rank-independent compiled graphs' -- but if the output is sometimes a `DTensor` and sometimes not (despite a `DTensor` input), that's surprising and worth one sentence of explanation. If it is always a `DTensor` under this configuration, drop the `isinstance` branch -- defensive code in a documentation example teaches the wrong lesson.

File: docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md [hunk 5]
```diff
@@ -217,6 +217,157 @@ with open("fn_with_model.pt", "rb") as f:
...
+        loss = local_out.sum()
+        loss.backward()
+        optimizer.step()
+        if rank == 0:
+            print(f"step {step}: loss = {loss.item():.4f}")
+
+    dist.destroy_process_group()
```
Comment (aorenste) [inline docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md:351]: Worth mentioning whether `save_compiled_function` / `load_compiled_function` works for artifacts compiled under `compile_on_one_rank=True`. The Training section shows save/load for the single-process case (lines 258-270) but the distributed section omits it entirely. If it works the same way, say so; if not, that's an important limitation to call out here.

**Review by aorenste — APPROVED** [review]
Approving w/ comments

File: docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md [hunk 6]
```diff
@@ -85,16 +87,16 @@ model = MyModel()
 # AOT compile the forward method.
 compiled_fn = torch.compile(
     model, fullgraph=True
```
Comment (aditvenk) [inline docs/source/user_guide/torch_compiler/torch.compiler_aot_compile.md:89]: Should this document change to `model.compile()` ? I believe the recommendation is to use that flow instead of torch.compile(model) ?

**Review by aditvenk — APPROVED** [review]
